intake_netflow/v9.py

```diff
@@ -0,0 +1,520 @@
+            self.records = payload
+
+    def __len__(self):
+        return s_type_length.size + len(self.records) * self.record_length
+
+    def __iter__(self):
+        return iter(self.records)
+
+    @staticmethod
+    def decode(source):
+        id, length = read_and_unpack(source, s_type_length)
+        payload = source.read(length - s_type_length.size)
+        return functools.partial(DataFlowSet, id, payload)
+
+    def encode(self):
+        raw = s_type_length.pack(self.template.id, len(self))
+        for record in self.records:
+            for field, value in zip(self.template.fields, record):
+                raw += field.struct.pack(value)
+        return raw
+
+
+def decode_flowset(source):
+    raw = source.peek(s_flowset.size)[:s_flowset.size]
+    flowset_id = s_flowset.unpack(raw)[0]
+    if flowset_id == 0:
+        return TemplateFlowSet.decode(source)
+    if flowset_id > 255:
+        return DataFlowSet.decode(source)
+    raise Exception("unknown flowset id '{}'".format(flowset_id))
+
+
+class ExportPacket(object):
+    def __init__(self, flowsets, header=None):
+        self.header = header if header else Header(count=len(flowsets))
+        self.flowsets = flowsets
+
+    def update_cache(self, cache):
+        """Update cache of template records."""
+        for flowset in self.flowsets:
+            if not isinstance(flowset, TemplateFlowSet):
+                continue
+            for id, record in flowset.templates.items():
+                cache[id] = record
+
+    def apply(self, templates):
+        """Deserialize partially-decoded data flowsets.
+
+        Deserialization of a data flowset is a two-step process because we
+        cannot assume the needed template is available when we encounter the
+        data flowset. Thus, we place the deserialization process on hold into
+        a packet is fully read. Then we re-scan the partially-decoded data
+        flowsets and finish deserialization.
+        """
+        for i, flowset in enumerate(self.flowsets):
+            if isinstance(flowset, functools.partial):
+                self.flowsets[i] = flowset(templates)
+
+    @staticmethod
+    def decode(source):
+        header = Header.decode(source)
+        flowsets = [decode_flowset(source) for _ in range(header.count)]
+        return ExportPacket(flowsets, header=header)
+
+    def encode(self):
+        raw = self.header.encode()
+        for flowset in self.flowsets:
+            if isinstance(flowset, functools.partial):
+                continue
+            raw += flowset.encode()
+        return raw
+
+
+class PacketStream(object):
+    def __init__(self, source):
+        """A read-only representation of serialized packets.
+
+        Parameters:
+            source : io.BufferedReader
+                Read-only input for packets.
+        """
+        self._source = source
+        self._cache = {}
+
+    def next(self):
+        if self._source.peek() == b'':
```

Comment:
Same comment as above, not sure you can rely on `peek`. Is this really the best way to check that a file has ended?